File: course_101/exercises/easy_7/lettercase_counter.rb
# Write a method that takes a string, and then returns a hash that contains 3 entries: one represents the number of characters in the string that are lowercase letters,
# one the number of characters that are uppercase letters, and one the number of characters that are neither.

def letter_case_count(string)
  case_count = Hash[:lowercase, 0, :uppercase, 0, :neither, 0]
  
  string.split('').each do |char|
    if char.match(/[a-z]/) then case_count[:lowercase] +=1
    elsif char.match(/[A-Z]/) then case_count[:uppercase] +=1
    else case_count[:neither] +=1
    end
  end
  case_count
end

puts letter_case_count('abCdef 123') == { lowercase: 5, uppercase: 1, neither: 4 }
puts letter_case_count('AbCd +Ef') == { lowercase: 3, uppercase: 3, neither: 2 }
puts letter_case_count('123') == { lowercase: 0, uppercase: 0, neither: 3 }
puts letter_case_count('') == { lowercase: 0, uppercase: 0, neither: 0 }

File: course_js210_computational_thinking_and_javascript_programming/exercises/js210_small_problems/string_and_text_processing/lettercase_counter.js
// Write a function that takes a string and returns an object containing three properties: one representing the number of characters in the string that are lowercase letters, one representing the number of characters that are uppercase letters, and one representing the number of characters that are neither.


// solution 1
function letterCaseCount(string) {
  let count = {
    lowercase: 0,
    uppercase: 0,
    neither: 0,
  };
  
  string.split('').forEach(function (character) {
    if (character.match(/[A-Z]/)) {
      count.uppercase += 1;
    } else if (character.match(/[a-z]/)) {
      count.lowercase += 1;
    } else {
      count.neither += 1;
    }
  });

  return count;
}

// solution 2
function letterCaseCount(string) {
  let lowerCaseCount = string.match(/[a-z]/g) || [];
  let upperCaseCount = string.match(/[A-Z]/g) || [];
  let neitherCaseCount = string.match(/[^a-z]/ig) || [];

  return {
    lowercase: lowerCaseCount.length,
    uppercase: upperCaseCount.length,
    neither: neitherCaseCount.length,
  }
}

console.log(letterCaseCount('abCdef 123'));  // { lowercase: 5, uppercase: 1, neither: 4 }
console.log(letterCaseCount('AbCd +Ef'));    // { lowercase: 3, uppercase: 3, neither: 2 }
console.log(letterCaseCount('123'));         // { lowercase: 0, uppercase: 0, neither: 3 }
console.log(letterCaseCount(''));            // { lowercase: 0, uppercase: 0, neither: 0 }